libs/base/Data/Colist.idr

```diff
@@ -0,0 +1,228 @@
+module Data.Colist
+
+import Data.Maybe
+import Data.List
+import public Data.Zippable
+
+%default total
+
+||| A possibly infinite list.
+public export
+data Colist : (a : Type) -> Type where
+  Nil : Colist a
+  (::) : a -> Inf (Colist a) -> Colist a
+
+--------------------------------------------------------------------------------
+--          Creating Colists
+--------------------------------------------------------------------------------
+
+||| Convert a list to a `Colist`.
+public export
+fromList : List a -> Colist a
+fromList []        = Nil
+fromList (x :: xs) = x :: fromList xs
+
+||| Convert a stream to a `Colist`.
+public export
+fromStream : Stream a -> Colist a
+fromStream (x :: xs) = x :: fromStream xs
+
+||| Create a `Colist` of only a single element.
+public export
+singleton : a -> Colist a
+singleton a = a :: Nil
+
+||| An infinite `Colist` of repetitions of the same element.
+public export
+repeat : a -> Colist a
+repeat v = v :: repeat v
+
+||| Create a `Colist` of `n` replications of the given element.
+public export
+replicate : Nat -> a -> Colist a
+replicate 0     _ = Nil
+replicate (S k) x = x :: replicate k x
+
+||| Produce a `Colist` by repeating a sequence.
+public export
+cycle : List a -> Colist a
+cycle Nil       = Nil
+cycle (x :: xs) = run x xs
+  where run : a -> List a -> Colist a
+        run v []        = v :: run x xs
+        run v (y :: ys) = v :: run y ys
+
+||| Generate an infinite `Colist` by repeatedly applying a function.
+public export
+iterate : (a -> a) -> a -> Colist a
+iterate f a = a :: iterate f (f a)
+
+||| Generate a `Colist` by repeatedly applying a function.
+||| This stops with `Nil` once the function returns `Nothing`.
+public export
+iterateMaybe : (f : a -> Maybe a) -> Maybe a -> Colist a
+iterateMaybe _ Nothing  = Nil
+iterateMaybe f (Just x) = x :: iterateMaybe f (f x)
+
+||| Generate an infinite `Colist` by repeatedly applying a function
+||| to a seed value.
+public export
+unfoldr : (f : s -> (s,a)) -> s -> Colist a
+unfoldr f s = let (s2,a) = f s
+               in a :: unfoldr f s2
+
+||| Generate an `Colist` by repeatedly applying a function
+||| to a seed value.
+||| This stops with `Nil` once the function returns `Nothing`.
+public export
+unfoldrMaybe : (f : s -> Maybe (s,a)) -> s -> Colist a
```

Comment:
This is the proper `unfold` for Colist.

Reply:
Renamed this one and wrote a similar `unfold` function for `Colist1`.